# gridss_filter
**gridss_filter** is a framework designed to assist in the prioritization of germline structural variants called by the [GRIDSS](https://github.com/PapenfussLab/gridss) software.

## Prerequisites 
- GRIDSS v2.13.2  
- RepeatMasker v4.1.5  
- R v4.4.2  
- R packages required (they will be installed while running the scripts if missing):  
  - CRAN packages: `assertthat`, `DiagrammeR`, `DiagrammeRsvg`, `dplyr`, `optparse`, `rsvg`, `stringr`, `yaml`  
  - Bioconductor packages: `GenomicRanges`
 

Input BAM files must be processed with GRIDSS and annotated using `gridss_annotate_vcf_repeatmasker`.

---

## Installation
To get started, clone this repository:

```bash
git clone https://github.com/emunte/gridss_filter.git
cd gridss_filter
```

---

## Running
### `merge_gridss_vcfs.R`
This script processes VCF files annotated with `gridss_annotate_vcf_repeatmasker` and extracts all variants with `PASS` filter. It generates two dataframes:

- **Variants with a single breakend**: entries that represent an isolated breakend.
- **Variants with two breakends**: paired breakends that are linked via `ID` and `MATEID` fields and belong to the same `EVENT`. These are merged into a single row representing the full structural variant.

The goal is to simplify downstream analysis by consolidating paired breakends into unified events and separating them from single breakend calls.

````
Rscript merge_gridss_vcfs.R --txt <paths_vcfs.txt> --output <output_folder> [--cores N]
````

#### Arguments

##### `--txt` (`-t`)
Path to a tab-delimited text file containing VCF paths and sample metadata. The file must have the following format:

| path  | sample | run | genes.interest |
|-----------|-----------|-----------|-----------|
| /path/to/sample1_vcf_repeatmasker.vcf | sample1  | run1  | gene1, gene2, gene3, gene4|
| /path/to/sample2_vcf_repeatmasker.vcf | sample2  | run1  | gene1, gene3, gene5|
| /path/to/sampleN_vcf_repeatmasker.vcf| sampleN  | runX  | geneY|

- `sample(N)` should match the exact sample name
- `genes.interest` should list comma-separated gene names

##### `--output` (`-o`):
Path to the output folder where result files will be stored.


##### `--cores` (`-c`):
The number of cores to use. 

#### Output
A `merge_gridss_vcfs` folder will be created inside the specified output directory. Two tab-separated files will be generated:

- `output_folder/merge_gridss_vcfs/merge_gridss_two_break.txt`: Variants involving two breakends that are linked through `ID` and `MATEID`.
- `output_folder/merge_gridss_vcfs/merge_gridss_one_break.txt`: Variants with a single breakend (unpaired).


### `filter_gridssR.R`

This script reads the two dataframes generated by merge_gridss_vcfs.R and applies different filters to prioritize germiline variants. 

```bash
Rscript filter_gridss.R --input <merge_gridss_output_folder> --bedFile <genes.bed> --params <params.yaml> --output <results_folder> --name <project_name> [--mergeVariants N]
````



